pkg/filesystem/nydus_filesystem.go

```diff
@@ -0,0 +1,138 @@
+// Copyright © 2021 Alibaba Group Holding Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package filesystem
+
+import (
+	"context"
+	"fmt"
+	"path/filepath"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/alibaba/sealer/logger"
+	"github.com/alibaba/sealer/pkg/env"
+
+	"github.com/alibaba/sealer/pkg/runtime"
+	v2 "github.com/alibaba/sealer/types/api/v2"
+
+	"github.com/alibaba/sealer/common"
+	"github.com/alibaba/sealer/utils"
+	"github.com/alibaba/sealer/utils/ssh"
+)
+
+type NydusFileSystem struct {
+	*FileSystem
+}
+
+func (c *NydusFileSystem) MountRootfs(cluster *v2.Cluster, hosts []string, initFlag bool) error {
+	clusterRootfsDir := common.DefaultTheClusterRootfsDir(cluster.Name)
+	//scp roofs to all Masters and Nodes,then do init.sh
+	md, err := runtime.LoadMetadata(common.DefaultMountCloudImageDir(cluster.Name))
+	if err != nil || md == nil {
+		return fmt.Errorf("LoadMetadata failed %v", err)
+	}
+	if md.NydusFlag {
+		nydusdfiledir := filepath.Join(common.DefaultMountCloudImageDir(cluster.Name), "nydusdfile")
+		nydusdcp := fmt.Sprintf("rm -rf %s && cp -r %s %s", common.DefaultNydusdfileDir, nydusdfiledir, common.DefaultNydusdfileDir)
+		_, err := utils.RunSimpleCmd(nydusdcp)
+		if err != nil {
+			return fmt.Errorf("cp nydusdfile failed %v", err)
+		}
+		if err := mountNydusRootfs(hosts, clusterRootfsDir, cluster, initFlag); err != nil {
+			return fmt.Errorf("mount rootfs failed %v", err)
+		}
+	} else {
+		if err := mountRootfs(hosts, clusterRootfsDir, cluster, initFlag); err != nil {
+			return fmt.Errorf("mount rootfs failed %v", err)
+		}
+	}
+	return nil
+}
+
+const (
+	NydusdMountUpper = "/var/lib/sealer/tmp/nydusdupper"
```

Comment:
Save nydus blob in tmp dir? I think /var/lib/sealer/data/<cluster_name>/blob/nydusdupper is better.  Because if has multiple cluster, tmp data will conflict I hink.